Add multitrack video dynamic bitrate support

lexano-ivs opened the pull request:
### Description

Dynamic bitrate (or DBR to keep it short), is a beta feature enabled via Settings -> Advanced -> Network -> "Dynamically change bitrate to manage congestion (Beta)" in the OBS GUI. This feature automatically modulates the video bitrate based on network bandwidth to avoid dropping frames. Support for DBR with multitrack video had been disabled in OBS while it was being tested in the TEB (Twitch Enhanced Broadcasting) beta program. This PR adds support for multitrack video dynamic bitrate.

Note that this PR depends on 2 previous PRs, due to a shared common function `HandleIncompatibleSettings()` that requires changes in all 3 PRs:

- https://github.com/obsproject/obs-studio/pull/12003
- https://github.com/obsproject/obs-studio/pull/12004

The dependent PRs were merged as of 2025.04.30 and this PR has been rebased against `master`.

### Motivation and Context
Multitrack video sends several video streams in parallel hence uses more overall bandwidth than sending a single stream. In some cases the streamers' available uplink bandwidth is not adequate for multitrack video, so modulating the bitrate on each stream is needed. 

### How Has This Been Tested?
Testing of multitrack DBR has been occurring for several months in the TEB beta program. In addition, it has been locally tested with NVIDIA GPUs, AMD GPUs (9070XT and 7800XT), and Intel Arc A770.

NVIDIA GPUs work well with multitrack DBR. 

AMD GPUs also showed issues with DBR, however this PR introduces changes that forces IDRs on the appropriate frame without needing to flush or reinitialize the AMF pipeline to solve the issues.

The Intel Arc A770 test revealed an issue:

```
15:12:49.701: [qsv encoder: 'multitrack video video encoder 1'] Failed to reconfigure
15:12:50.009: [qsv encoder: 'multitrack video video encoder 0'] Failed to reconfigure
```

The failure to reconfigure resulted in all streams ceasing to transmit data and an eventual hang of OBS which needed to be terminated with Task Manager. The multitrack DBR feature checks if the encoder supports reconfiguration (it looks for `OBS_ENCODER_CAP_DYN_BITRATE`. It is possible that Intel Arc GPUs do not support reconfiguration but have the bit set. This requires further investigation and testing. 

2025.08.08 Update: The PR has been updated to not invoke DBR if the encoder is Intel QSV. The OBS hang issue is no longer happening with the code available at https://github.com/obsproject/obs-studio/pull/12481, however the IDRs are not aligned across the segments. Once the issue is identified and fixed, reverting the last commit in this PR should be only change needed to enable multitrack DBR for Intel.

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [ ] I have included updates to all appropriate documentation.

palana committed:
libobs: Refactor `encoder_group` access

Also reconfigure all grouped encoders on the same frame.

palana committed:
obs-outputs: Implement multitrack dynamic bitrate

palana committed:
frontend: Add dynamic bitrate support to multitrack video

The bitrate interpolation points are sent in JSON configuration and
need to be forwarded to `create_video_encoders()`. Also remove the
`HandleIncompatibleSettings()` handler because there are no longer any
incompatible settings.

Commit:
obs-ffmpeg: Rework AMF dynamic bitrate change behavior

The `amf_xxx_update()` functions for AVC, HEVC, and AV1 are
unconditionally invoking Flush() and ReInit() even though this is not
required in all situations. Changing the bitrate at least in CBR mode
does not require the flush operation and can result in unaligned IDR
frames across members of an encoder group. Do not flush the pipeline
for CBR bitrate changes; instead, force an IDR to occur with the
bitrate change.

RytoEX reviewed (changes requested):
This can be rebased now that #12003 and #12004 are merged. Remove the changes to all locale files except `en-US.ini`, as the others will be written/overwritten when we pull in translations.

Commit:
frontend: Remove multitrack locale strings

Remove the locale strings that are no longer needed now that
`HandleIncompatibleSettings()` is removed, and replace some
`<br>` tags with \n to fix rendering issues on MacOS.

lexano-ivs commented:
> This can be rebased now that #12003 and #12004 are merged. Remove the changes to all locale files except `en-US.ini`, as the others will be written/overwritten when we pull in translations.

Thanks for the tip re: locale files. Changes have been made and rebased to `master`.

Commit:
obs-outputs: Use multitrack DBR capability flag

Use the new `OBS_ENCODER_CAP_MULTITRACK_DYN_BITRATE` flag when
streaming with multiple tracks, instead of relying on the single-
track `OBS_ENCODER_CAP_DYN_BITRATE` flag. In multitrack cases where
multiple codecs are being used (e.g. mixed AVC and HEVC), all encoders
must support the multitrack capability flag otherwise the dynamic
bitrate (DBR) feature is disabled for the streaming session.

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1228,6 +1229,78 @@ static int try_connect(struct rtmp_stream *stream)
+		if (all_array_size != size) {
+			warn("size %zu of bitrate interpolation data for "
+			     "encoder '%s' does not match overall size %zu",
+			     size, name, all_array_size);
```
Unwrap and rewrap. Column limit is 120. Generally, logged strings should be unbroken.

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1266,55 +1338,91 @@ static bool init_connect(struct rtmp_stream *stream)
+				     "The encoder '%s' feature is currently disabled pending a fix for Intel QSV DBR support.",
+				     obs_encoder_get_name(enc));
+			}
+			continue;
```
I would almost rather disable Dynamic Bitrate support in the QSV encoder and rely on the feature flag than have the RTMP stream output be aware of specific encoder quirks. That, or introduce a new encoder capability flag for "multitrack video dynamic bitrate" (though this means encoders are aware of some specific output/protocol feature/implementation) or "forced IDR recreation/alignment". Is there any prior art of manually adjusting per encoder in this or other outputs?

lexano-ivs replied:
There is no prior art in any of the outputs that I can find where the behavior is adjusted based on the encoder identification. The intention with the current `encoder_id` based solution is a short-term fix until the QSV encoder supports IDR alignment across renditions and codecs (currently AVC and HEVC). 

We had considered keying off the `OBS_ENCODER_CAP_DYN_BITRATE` alone, but this would mean disabling it globally for QSV encoders, so the single-stream case would also stop working. 

My original thought back when this all started was to introduce a new capability flag similar to what you suggest, something like `OBS_ENCODER_CAP_MULTITRACK_DBR`, but taking this route also likely means the new flag is there forever, and the output (RTMP in the current case) needs to be aware of multitrack vs non-multitrack, which is not a problem, just adds more decision logic. I'll try this locally.

lexano-ivs replied:
@RytoEX I've removed the commit with the string-based `encoder_id` solution for identifying QSV encoders. The new `OBS_ENCODER_CAP_MULTITRACK_DYN_BITRATE` flag is the preferred method.

Commit:
libobs: Add multitrack dynamic bitrate capability

The `OBS_ENCODER_CAP_DYN_BITRATE` flag indicates if an encoder is
capable of supporting dynamic bitrate changes without disturbing
the stream. In the case of multitrack video, dynamic bitrate changes
are also possible, however the encoder must be able to change bitrates
for multiple renditions and multiple codecs and maintain IDR alignment.
Add the `OBS_ENCODER_CAP_MULTITRACK_DYN_BITRATE` to specify this
capability for the encoder.

Commit:
plugins: Apply multitrack dynamic bitrate flag

Enable the `OBS_ENCODER_CAP_MULTITRACK_DYN_BITRATE` capability on
the encoders that support it.

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1266,55 +1334,103 @@ static bool init_connect(struct rtmp_stream *stream)
+		bool has_multitrack_dbr_cap = (caps & OBS_ENCODER_CAP_MULTITRACK_DYN_BITRATE) != 0;
+
+		// For multitrack, all encoders must support multitrack DBR. For single track, encoder must support
+		// standard DBR.
```
I usually prefer just spelling out acronyms, or at least defining it in text:
```suggestion
		// For multitrack, all encoders must support multitrack Dynamic Bitrate. For single track, encoder must
		// support standard Dynamic Bitrate.
```

or:

```suggestion
		// For multitrack, all encoders must support multitrack Dynamic Bitrate (DBR). For single track, encoder
		// must support standard Dynamic Bitrate.
```

lexano-ivs replied:
I've updated the comment in the latest push.

RytoEX reviewed (changes requested):
Nit. Most of this was previously reviewed.

PatTheMav commented on libobs/obs-encoder.c:
```diff
@@ -346,6 +342,26 @@ static void maybe_set_up_gpu_rescale(struct obs_encoder *encoder)
 	pthread_mutex_unlock(&obs->video.mixes_mutex);
 }
 
+static uint32_t gcd(uint32_t a, uint32_t b)
```
Per our code style guidelines:

> Do not use specific integer types outside of binary protocols - for the vast majority of cases using an “int” is sufficient and using a signed type also prevents unexpected overflow issues where a small negative number becomes a massive positive number. Use sized types (and unsigned types) when interacting with binary protocols, bitfields (e.g. to explicitly remove any special meaning from the most-significant bit). Prefer 64-bit integers over unsigned 32-bit integers for large numbers.

Admittedly we updated those guidelines after this PR was written, but the rule has been an established industry practice already.

See also: [Unsigned integers, and why to avoid them ](https://www.learncpp.com/cpp-tutorial/unsigned-integers-and-why-to-avoid-them/).

lexano-ivs replied:
Update to use `int` with the latest push.

PatTheMav commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1538,15 +1655,60 @@ static bool dbr_bitrate_lowered(struct rtmp_stream *stream)
 
+#ifndef min
+#define min(a, b) (((a) < (b)) ? (a) : (b))
+#endif
```
Macros are evil and function-like macros are evil squared. Given that this is used only once in the function body, replace the occurrence with the actual code.

lexano-ivs replied:
Evil macro removed and replaced with actual code in the one occurrence of usage.

PatTheMav reviewed (commented):
When using custom RTMP output with a custom JSON file, the current code will emit the following error when that JSON does not contain encoder-specific settings:

```
error: obs-data.c: [obs_data_create_from_json] Failed reading json string (1): '[' or '{' expected near 'null'
```

This is caused by `OBSEncoderAutoRelease create_video_encoder` calling `obs_data_create_from_json` without checking whether the result of `encoder_config.settings.dump().c_str()` actually yields a non-empty string.

PatTheMav commented:
This also seems to introduce a breaking change: A multitrack configuration that I've used to test custom rtmp output is not accepted anymore and output is aborted, e.g. because my config does not contain explicit audio encoder information.

Is this breaking change intended? If so we'd need to introduce versioning and proper migration documentation of users who might use this feature right now.

Because I at least recall being able to successfully set up a stream just with a JSON that contains two `encoder_configurations` objects, an `ingest_endpoints` and a `meta` object.

In my testing it also seems that the code merges an existing custom configuration from the "custom RTMP" service with Twitch's downloaded configuration, which makes the custom config overwrite existing fields: `audio_configurations` became empty, `encoder_configurations` was replaced (so it's not even a "deep" merge).

The problem is that this is nonsensical for the user: They cannot see the custom config until they switch from Twitch to the custom service, but even then they have to delete a custom config they might want to use with their custom service just to make Twitch work (again).

PatTheMav commented:
<img width="372" height="368" alt="Bildschirmfoto 2026-05-22 um 17 20 29" src="https://github.com/user-attachments/assets/fc1b6d74-8d47-4cb9-8e3d-817a1f3f2102" />

The error message also contains raw HTML tokens.

The follow-up dialog (itself not particularly helpful to trigger two individual dialogs one after another, they should be merged, but OBS' codebase uses this "check each thing individually and show error dialogs individually" pattern elsewhere) also mentions NVENC and AMD and their respective drivers, which doesn't make much sense on macOS.

PatTheMav commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1228,6 +1229,74 @@ static int try_connect(struct rtmp_stream *stream)
 	return init_send(stream);
 }
 
+static bool build_dbr_interpolation_table(struct rtmp_stream *stream, const char *interpolation_data)
```
I'm trying to understand the purpose of this code, but as far as I can tell it effectively transposes the bitrate table from each row representing an encoder and each column representing a quality level to each row representing a quality level, with each column representing the encoder (and the index inside the row has to match the index inside the encoder list)?

lexano-ivs replied:
I added a description in the comments showing the mapping. Yes it's a transpose function. The new comments should spell it out clearly.

PatTheMav replied:
Oh that's much appreciated.

PatTheMav commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1538,15 +1655,60 @@ static bool dbr_bitrate_lowered(struct rtmp_stream *stream)
+		return;
+
+	size_t dbr_base_column = stream->dbr_interpolation_table.num - 1;
+	for (size_t column = 0; column < stream->dbr_interpolation_table.num; column++) {
```
```suggestion
	for (size_t column = 0; column < stream->dbr_interpolation_table.num; ++column) {
```
If post-increment semantics are not required, use pre-increment by default, applies throughout.

lexano-ivs replied:
Fixed

PatTheMav commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1538,15 +1655,60 @@ static bool dbr_bitrate_lowered(struct rtmp_stream *stream)
+
+	struct dbr_interpolation_point *dbr_base_point = &stream->dbr_interpolation_table.array[dbr_base_column];
+	struct dbr_interpolation_point *dbr_upper_point = &stream->dbr_interpolation_table.array[dbr_upper_column];
+	long deltas[MAX_OUTPUT_VIDEO_ENCODERS] = {0};
```
Any particular reason why this code uses `long` instead of `int`?

lexano-ivs replied:
`long deltas[]`, `long remaining_bitrate`, etc. — matches the existing `struct dbr_interpolation_point` field type (`long bitrates[MAX_OUTPUT_VIDEO_ENCODERS]`) and all other DBR variables (`dbr_cur_bitrate`, `dbr_orig_bitrate`, `dbr_est_bitrate`, `dbr_inc_bitrate`, `audio_bitrate`) which are `long` in `rtmp-stream.h`.

PatTheMav commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1266,55 +1334,103 @@ static bool init_connect(struct rtmp_stream *stream)
-	obs_encoder_t *aenc = obs_output_get_audio_encoder(stream->output, 0);
-	obs_data_t *vsettings = obs_encoder_get_settings(venc);
-	obs_data_t *asettings = obs_encoder_get_settings(aenc);
+	long long overall_audio_bitrate = 0;
```
Why `long long` here and below?

lexano-ivs replied:
`long long overall_audio_bitrate` / `overall_video_bitrate` — matches the return type of `obs_data_get_int()` which returns `long long` (see `libobs/obs-data.h:121`)

PatTheMav commented on libobs/obs-encoder.c:
```diff
@@ -366,6 +382,12 @@ static void add_connection(struct obs_encoder *encoder)
 
 	if (encoder->encoder_group) {
 		pthread_mutex_lock(&encoder->encoder_group->mutex);
+		if (encoder->encoder_group->frame_rate_divisors_lcm == 0)
+			encoder->encoder_group->frame_rate_divisors_lcm = 1;
+
+		encoder->encoder_group->frame_rate_divisors_lcm =
+			lcm(encoder->encoder_group->frame_rate_divisors_lcm, encoder->frame_rate_divisor);
```
How is this new "lowest common denominator" of all frame rate divisors connected to the dynamic bitrate feature?

lexano-ivs replied:
Added a comment section in the code to explain.

Commit:
frontend: Fix null settings and DBR for custom configs

When using custom RTMP output with a JSON config that omits
encoder settings, settings.dump() produces "null" which
obs_data_create_from_json cannot parse. Fall back to
obs_data_create() for null/non-object settings so encoders
use their defaults.

Additionally, when bitrate_interpolation_points is absent,
the code unconditionally set interpolation_table_data as
empty arrays. This caused build_dbr_interpolation_table to
wipe the default linear table, leaving DBR enabled but
unable to adjust bitrates during congestion. Only set the
interpolation data when all encoders provide it, otherwise
fall back to the default linear 0-to-max interpolation.

PatTheMav commented:
@lexano-ivs as discussed I've went back in and tried to make sense of the custom rtmp situation. I've added #13487 to simplify the logic in current code which should also ensure that any existing "custom configuration" a user might have are not applied to any non-custom service.

The response schema used by `GetClientConfiguration` is the source of truth, so any custom JSON file used for a custom RTMP service needs to have the same required fields, we don't check the file explicitly though. 

"Custom RTMP" with multitrack enabled remains in "here be dragons" territory, but at least with that PR any user mucking around with it won't negatively impact functionality of any non-custom service that supports multitrack.

So there is no regression with regards to custom configurations.

lexano-ivs commented:
> @lexano-ivs as discussed I've went back in and tried to make sense of the custom rtmp situation. I've added #13487 to simplify the logic in current code which should also ensure that any existing "custom configuration" a user might have are not applied to any non-custom service.
> 
> The response schema used by `GetClientConfiguration` is the source of truth, so any custom JSON file used for a custom RTMP service needs to have the same required fields, we don't check the file explicitly though.
> 
> "Custom RTMP" with multitrack enabled remains in "here be dragons" territory, but at least with that PR any user mucking around with it won't negatively impact functionality of any non-custom service that supports multitrack.
> 
> So there is no regression with regards to custom configurations.

Thank you. I just pushed some changes (still building as I type) that includes some protection for bad configs, specifically commit 84364883d93632fdc47dc7f272ccf6676a0a4cc9. I'll check if #13487 conflicts with my change.

lexano-ivs commented:
> When using custom RTMP output with a custom JSON file, the current code will emit the following error when that JSON does not contain encoder-specific settings:
> 
> ```
> error: obs-data.c: [obs_data_create_from_json] Failed reading json string (1): '[' or '{' expected near 'null'
> ```
> 
> This is caused by `OBSEncoderAutoRelease create_video_encoder` calling `obs_data_create_from_json` without checking whether the result of `encoder_config.settings.dump().c_str()` actually yields a non-empty string.

Should be fixed with latest push, commit 84364883d93632fdc47dc7f272ccf6676a0a4cc9

lexano-ivs commented:
> <img alt="Bildschirmfoto 2026-05-22 um 17 20 29" width="372" height="368" src="https://private-user-images.githubusercontent.com/9436503/596849269-fc1b6d74-8d47-4cb9-8e3d-817a1f3f2102.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Nzk5MDk5NjksIm5iZiI6MTc3OTkwOTY2OSwicGF0aCI6Ii85NDM2NTAzLzU5Njg0OTI2OS1mYzFiNmQ3NC04ZDQ3LTRjYjktOGUzZC04MTdhMWYzZjIxMDIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDUyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA1MjdUMTkyMTA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmM2ZDQ4MzA0ZWQ1MDNmYzJlYWIzM2E2NGNiMTQzMDUxNGNiOTA1NDEzMDVhYmQ5ODE4ZDk2ZjVlMjRhYzFmZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmcmVzcG9uc2UtY29udGVudC10eXBlPWltYWdlJTJGcG5nIn0.4X6QltZo8XXCxVPnplrvf6nZUaE6ZBcwJOaPolGJQh4">
> The error message also contains raw HTML tokens.
> 
> The follow-up dialog (itself not particularly helpful to trigger two individual dialogs one after another, they should be merged, but OBS' codebase uses this "check each thing individually and show error dialogs individually" pattern elsewhere) also mentions NVENC and AMD and their respective drivers, which doesn't make much sense on macOS.

I removed the "<br>" entries in the Multitrack-realated en-US.ini file with the latest push. This should fix the rendering in the dialog box on MacOS. Please test if you can.

PatTheMav commented on frontend/utility/MultitrackVideoOutput.cpp:
```diff
@@ -684,6 +618,12 @@ static bool create_video_encoders(const GoLiveApi::Config &go_live_config,
 		if (recording_output)
 			obs_output_set_video_encoder2(recording_output, encoder, i);
+
+		auto &data = go_live_config.encoder_configurations[i].bitrate_interpolation_points;
+		if (data.has_value())
+			bitrate_interpolation_array.push_back(*data);
+		else
+			bitrate_interpolation_array.push_back(json::array());
```
Last remaining nit (I hope): For C++ code we require all conditional and loop blocks to be enclosed with curly braces. Probably applies to other changes in the file as well.

lexano-ivs replied:
I added braces where needed in `frontend/utility/MultitrackVideoOutput.cpp` and `plugins/obs-ffmpeg/texture-amf.cpp`, in the areas within PR scope.

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1540,13 +1683,55 @@ static bool dbr_bitrate_lowered(struct rtmp_stream *stream)
+		overall_delta += deltas[i];
+		remaining_bitrate -= dbr_base_point->bitrates[i];
+	}
+
+	if (overall_delta < 1)
+		overall_delta = 1;
+	double ratio = remaining_bitrate / (double)overall_delta;
```
While we do not require braces in C for singe-line `if` statements, we do recommend a blank line after them instead.

lexano-ivs replied:
I added a blank line.

RytoEX reviewed (changes requested):
Nit.

RytoEX reviewed (changes requested):
Hopefully final asks:
* Squash the fixup commit into the originating commit(s)
* Rebase

lexano-ivs commented:
> Hopefully final asks:
> 
> * Squash the fixup commit into the originating commit(s)
> * Rebase

Done and done.
